### PR TITLE
Initial code

### PR DESCRIPTION
**Pull request by jhford**
@jonasfj I haven't actually run this quite yet, just wanted to get your opinions...  I am thinking of adding something to the wrapper to let you run a given command on only a subset of regions, so you could use the same object to run on individual regions as well as on all configured

### CODE REVIEW

**Comment by jhford** [comment]
I'd like to be able to let a specific request to be able to be run on only a subset of regions configured, but I'm not sure how I should make the syntax for that.

**Comment by jonasfj** [comment]
@jhford, Why not just create an instance of aws-sdk-promise for each region.
I think that is the simple thing to do... Notice that you only operate on multiple regions when fecthing spot price history... For all other operation you want to specify a region...

**Comment by jonasfj** [comment]
I would think that having execute the same request in multiple regions is very much a corner case...

Ie a construct like:
```js
var ec2 = {
  'us-west-2':   new aws.EC2(_.defaults({region: 'us-west-2'}, config)),
  ....
};
```
Would do fine... As you would always be operating on one region at the time. Except for when you're downloading spot price history.

you can even make this as
```js
var ec2 = {};
['us-west-2', ...].forEach(function(region) {
  ec2[region] = new aws.EC2(_.defaults({region: region}, config));
});
```

**Comment by jhford** [comment]
It's a corner case until you want to write a provisioner which is multi-region capable. Every single call other than creating and killing instances needs to be done in every single region.  I actually have written in support for doing something like .inRegion('us-west-2').

Using a plain object to store all the objects still means that for each EC2 call, I still have to gather and sort the results by region.  This library has tests and makes that trivial.

```javascript
var multiAws = multiAws = new aws(aws.EC2, {maxRetries: 5}, ['us-west-2', 'us-west-1'])
var p = multiAws.describeSpotPriceHistory(req);
p = p.then(multiAws.requestSpotInstance.inRegion('us-west-2', req2));
```
If I had just a dictionary of ec2 objects, I would have to do something like
```javascript
var ec2 = {};
['us-west-2', ...].forEach(function(region) {
  ec2[region] = new aws.EC2(_.defaults({region: region}, config));
});

var p = Promise.all([
    ec2['us-west-1'].describeSpotPriceHistory(req).promise(),
    ec2['us-west-2'].describeSpotPriceHistory(req).promise(),
]);
p = p.then(function(res) {
    //do stuff here to gather all the results
    return Promise.resolve(<something that was created>);
});
p = p.then(ec2['us-west-2'].requestSpotInstance(req2).promise();
```

I'd rather do the first :)

File: index.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
+'use strict';
+
+var aws = require('aws-sdk-promise');
+var lodash = require('lodash');
+var debug = require('debug')('multi-region-promised-aws');
+var Promise = require('promise');
+
+// TODO: allow for each region to have its own defaults and settings
+function MultiAWS(nameOrConstructor, config, regions) {
```
Comment (jonasfj) [inline index.js]: Please add a doc comment like:
```
 /**
 * This class represents... bla.. .bla...
 *
 * config:  // AWS config...
 * etc....
 */
```

Anything like that... look at other code we have base.Entity has a lot of the comments in a consistent format.
We don't have docs generator for them (sadly) maybe one day we'll find something that actually works...

File: index.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    throw new Error('First argument must be a string or function');
+  }
+  
+  if (config.region) {
+    throw new Error('To avoid a footgun, your AWS config should not specify region');
+  }
+
+  // All I care that regions has a forEach, map and filter.  I don't
+  // care if it will return true for Array.isArray()
+  if (!regions.forEach || !regions.map || !regions.filter) {
```
Comment (jonasfj) [inline index.js]: This is `node` I usually trusts `instanceof Array`, isArray is probably better this is also okay... Just saying...

File: index.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    var p = Promise.all(regionPromises)
+
+    // Put everything into an region-keyed object
+    p = p.then(function(res) {
+      debug('all promises done');
+      var result = {};
+      regions.forEach(function(region, idx) {
+        result[region] = res[idx].data;
+      });
+      return Promise.resolve(result);
```
Comment (jonasfj) [inline index.js]: `Promise.resolve` not needed...

File: index.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    var regionPromises = regions.map(function(region) {
+      var apiObj = that.apiObjs[region];
+      return apiObj[name].apply(apiObj, args).promise(); 
+    });
+
+    // Make the request for each region
+    var p = Promise.all(regionPromises)
+
+    // Put everything into an region-keyed object
+    p = p.then(function(res) {
```
Comment (jonasfj) [inline index.js]: Nit but usually I would structure this as:
```js
return Promise.all(regionPromises).then(function(res) {
  ....
});
```

In fact I would just do:
```js
return Promise.all(regions.map(function(region) {
   ....
})).then(function(res) {
  ....
});
```

The intermediary `p` does improve readability if you have a long chain.. Here it's maybe over kill...

File: index.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    };
+
+    that[name].inRegion = function(region) {
+      var args = Array.prototype.slice.call(arguments, 1);
+      
+      var p = runApiMethod(name, [region], args);
+      
+      p = p.then(function(result) {
+        var thisRegions = result[region];
+        return Promise.resolve(thisRegions);
```
Comment (jonasfj) [inline index.js]: Again `Promise.resolve` is not needed...
Also I would just chain these together as:
```js
var args = Array.prototype.slice.call(arguments, 1);
return runApiMethod(name, [region], args).then(function(result) {
  return result[region];
});
```

Though you can probably do even simpler... I'm not sure what you need to use `runApiMethod` for in this case.

File: test/mocha.opts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+--ui tdd
+--timeout 30s
+--reporter tap
```
Comment (jonasfj) [inline test/mocha.opts]: So just FYI; everywhere we're using `spec` just saying...

File: test/runtests.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash -ve
+# USAGE: Run this file using `npm test` (must run from repository root)
+
+# Run tests
+mocha --timeout 15s
```
Comment (jonasfj) [inline test/runtests.sh]: I recommend again using both `mocha.opts` and options in `runtests.sh`.

**Comment by jonasfj** [comment]
Only minor comments, feel free to merge or address the comments... I suggest you fix some of the ones with promises.... So you get that right...  Do some chaining... etc...

File: index.js (hunk 5 above)
Comment (jhford) [inline index.js]: `runApiMethod` also strips the raw repsonse out, I don't want to change the behaviour of the underlying system depending on whether the call was made in one region or more as much as I can.

**Comment by jhford** [comment]
I addressed the comments and merged.

**Pull request merged**